### ISSUE TITLE
Compilation fails with GCC 6.2.0

### ISSUE THREAD

**Issue by PranayShah** [kaldi-asr/kaldi#1432]
I could not get both tools/ and src/ to compile with the following:
`uname -a`
`Linux asdf-VirtualBox 4.8.0-37-generic #39-Ubuntu SMP Thu Jan 26 02:27:07 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`
`gcc -v`
```
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/6/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 6.2.0-5ubuntu12' --with-bugurl=file:///usr/share/doc/gcc-6/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-6 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-6-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-6-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-6-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 6.2.0 20161005 (Ubuntu 6.2.0-5ubuntu12) 
```
Error thrown is:
```
/home/asdf/kaldi/tools/openfst/include/fst/arc-map.h: In member function ‘fst::GallicToNewSymbolsMapper<A, G>::ToArc fst::GallicToNewSymbolsMapper<A, G>::operator()(const FromArc&)’:
/home/asdf/kaldi/tools/openfst/include/fst/arc-map.h:968:14: error: inconsistent deduction for ‘auto’: ‘int’ and then ‘auto’
         for (auto i = 0, p = state_; i < w1.Size();
```
And similar errors at multiple places.
However compiling with version `4.9` does work.

**Comment by danpovey** [comment]
I found a reference to this error message here:

https://bugs.llvm.org//show_bug.cgi?id=30438

and the thread seems to indicate that it's a bug in that version of gcc
that has since been fixed.   Just don't use that version of gcc.

**Comment by jtrmal** [comment]
I reported this to openfst people as well, Kyle did some fixes which are
already in openfst 1.6.1 and was about to report it to GCC, but they alread
knew about that -- it will be fixed in gcc 6.4
Gcc 6.3 is broken as well. Use the 5.x GCC.
y.
